libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+                if (runner instanceof ShellOperationRunner) {
+                  const buildCacheContext: IOperationBuildCacheContext =
+                    this._getBuildCacheContextByRunnerOrThrow(runner);
+                  buildCacheContext.cobuildClusterId = cobuildClusterId;
+                }
+              }
+            }
+          }
+        }
+
+        return operations;
+      }
+    );
+
+    hooks.afterExecuteOperation.tapPromise(
+      PLUGIN_NAME,
+      async (runnerContext: IOperationRunnerContext): Promise<void> => {
+        const { runner, status, consumers, changedProjectsOnly } = runnerContext;
+        const buildCacheContext: IOperationBuildCacheContext | undefined =
+          this._getBuildCacheContextByRunner(runner);
+
+        let blockCacheWrite: boolean = !buildCacheContext?.isCacheWriteAllowed;
+        let blockSkip: boolean = !buildCacheContext?.isSkipAllowed;
+
+        switch (status) {
+          case OperationStatus.Skipped: {
+            // Skipping means cannot guarantee integrity, so prevent cache writes in dependents.
+            blockCacheWrite = true;
+            break;
+          }
+
+          case OperationStatus.SuccessWithWarning:
+          case OperationStatus.Success: {
+            // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+            blockSkip ||= !changedProjectsOnly;
+            break;
+          }
+        }
+
+        // Apply status changes to direct dependents
+        for (const item of consumers) {
+          const itemRunnerBuildCacheContext: IOperationBuildCacheContext | undefined =
+            this._getBuildCacheContextByRunner(item.runner);
+          if (itemRunnerBuildCacheContext) {
+            if (blockCacheWrite) {
+              itemRunnerBuildCacheContext.isCacheWriteAllowed = false;
+            }
+            if (blockSkip) {
+              itemRunnerBuildCacheContext.isSkipAllowed = false;
+            }
+          }
+        }
+      }
+    );
+
+    hooks.afterExecuteOperations.tapPromise(PLUGIN_NAME, async () => {
+      for (const { buildCacheProjectLogWritable } of this._buildCacheContextByOperationRunner.values()) {
+        buildCacheProjectLogWritable?.close();
+      }
+      this._buildCacheContextByOperationRunner.clear();
+    });
+  }
+
+  private _applyShellOperationRunner(runner: ShellOperationRunner, context: ICreateOperationsContext): void {
+    const {
+      buildCacheConfiguration,
+      cobuildConfiguration,
+      phaseSelection: selectedPhases,
+      projectChangeAnalyzer
+    } = context;
+    const { hooks } = runner;
+
+    const buildCacheContext: IOperationBuildCacheContext = this._getBuildCacheContextByRunnerOrThrow(runner);
+
+    hooks.beforeExecute.tapPromise(
+      PLUGIN_NAME,
+      async (beforeExecuteContext: IOperationRunnerBeforeExecuteContext) => {
+        const earlyReturnStatus: OperationStatus | undefined = await (async () => {
```

Comment:
Could you hoist this function definition and take the values that come from the parent scopes as arguments? I think it would make the file more readable.